File: 045. Jump Game II.scala
/*
Given an array of non-negative integers, you are initially positioned at the first index of the array.

Each element in the array represents your maximum jump length at that position.

Your goal is to reach the last index in the minimum number of jumps.

For example:
Given array A = [2,3,1,1,4]

The minimum number of jumps to reach the last index is 2. (Jump 1 step from index 0 to 1, then 3 steps to the last index.)

Note:
You can assume that you can always reach the last index.


*/
///注意28行的条件
object Solution {
    def jump(nums: Array[Int]): Int = {
        var n = nums.length;
        var mark = new Array[Int](n);
        var i = n-2;
        while(i>=0){
            mark(i) = Int.MaxValue;
            var end = math.min(i+nums(i),n-1)
            for(j<- i+1 to end){
                if(j==(n-1) || (nums(j)!=0 && mark(j)!=Int.MaxValue)){
                    mark(i) = math.min(1+mark(j),mark(i))
                }
            }
            i -= 1;
        }
        return mark(0)
    }
}

    val arr = Array(2,3,1,1,4)
    jump(arr)

    var arr = Array(5,9,3,2,1,0,2,3,3,1,0,0)
    jump(arr)

File: 045. Jump Game II.py
class Solution(object):
    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if  nums is None or len(nums)<=1:
            return 0
        res = 0
        pos = 0
        while pos<=len(nums)-1:
            if pos + nums[pos]>= len(nums)-1:
                return res + 1
            i = 1
            for k in range(2,min(nums[pos]+1,len(nums)-1)):
                if k+nums[pos+k]>i+nums[pos+i]:
                    i = k
            pos += i
            res += 1
            print(i)
        return res
            
